sjsonnet/src/sjsonnet/Importer.scala

```diff
@@ -1,37 +1,172 @@
+  }
+}
+
+case class FileParserInput(file: File) extends ParserInput {
+
+  private[this] val bufferedFile = new BufferedRandomAccessFile(file.getAbsolutePath, 1024 * 8)
+
+  private lazy val fileLength = file.length.toInt
+
+  override def apply(index: Int): Char = {
+    bufferedFile.readChar(index)
+  }
+
+  override def dropBuffer(index: Int): Unit = {}
+
+  override def slice(from: Int, until: Int): String = {
+    bufferedFile.readString(from, until)
+  }
+
+  override def length: Int = fileLength
+
+  override def innerLength: Int = length
+
+  override def isReachable(index: Int): Boolean = index < length
+
+  override def checkTraceable(): Unit = {}
+
+  private[this] lazy val lineNumberLookup: Array[Int] = {
+    val lines = mutable.ArrayBuffer[Int]()
+    val bufferedStream = new BufferedInputStream(new FileInputStream(file))
+    var byteRead: Int = 0
+    var currentPosition = 0
+
+    while ({ byteRead = bufferedStream.read(); byteRead != -1 }) {
+      if (byteRead == '\n') {
+        lines += currentPosition + 1
+      }
+      currentPosition += 1
+    }
+
+    bufferedStream.close()
+
+    lines.toArray
+  }
+
+  def prettyIndex(index: Int): String = {
+    val line = lineNumberLookup.indexWhere(_ > index) match {
+      case -1 => lineNumberLookup.length - 1
+      case n => math.max(0, n - 1)
+    }
+    val col = index - lineNumberLookup(line)
+    s"${line + 1}:${col + 1}"
   }
 }
 
+class BufferedRandomAccessFile(fileName: String, bufferSize: Int) {
+
+  // The file is opened in read-only mode
+  private val file = new RandomAccessFile(fileName, "r")
+
+  private val buffer = new Array[Byte](bufferSize)
+
+  private var bufferStart: Long = -1
+
+  private var bufferEnd: Long = -1
+
+  private val fileLength: Long = file.length()
+
+  private def fillBuffer(position: Long): Unit = {
+    file.seek(position)
+    val bytesRead = file.read(buffer, 0, bufferSize)
+    bufferStart = position
+    bufferEnd = position + bytesRead
+  }
+
+  def readChar(index: Long): Char = {
+    if (index >= fileLength) {
+      throw new IndexOutOfBoundsException(s"Index $index is out of bounds for file of length $fileLength")
+    }
+    if (index < bufferStart || index >= bufferEnd) {
+      fillBuffer(index)
+    }
+    buffer((index - bufferStart).toInt).toChar
+  }
+
+  def readString(from: Long, until: Long): String = {
+    if (!(from < fileLength && until <= fileLength && from <= until)) {
+      throw new IndexOutOfBoundsException(s"Invalid range: $from-$until for file of length $fileLength")
+    }
+    val length = (until - from).toInt
+    val stringBytes = new Array[Byte](length)
+
+    if (from >= bufferStart && until <= bufferEnd) {
+      // Range is within the buffer
+      System.arraycopy(buffer, (from - bufferStart).toInt, stringBytes, 0, length)
+    } else {
+      // Range is outside the buffer
+      file.seek(from)
+      file.readFully(stringBytes, 0, length)
+    }
+
+    new String(stringBytes)
```

Comment:
The current implementation uses `StandardCharsets.UTF_8`, so we should probably follow suit

Reply:
Done